addons/cloud_storage_azure/models/cloud_storage_azure.py

```diff
@@ -0,0 +1,134 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import re
+import requests
+from datetime import datetime, timedelta, date
+from urllib.parse import unquote, quote
+
+from odoo import models, _
+from odoo.exceptions import ValidationError
+from odoo.tools import ormcache
+
+from ..utils.cloud_storage_azure_utils import generate_blob_sas, get_user_delegation_key
+
+
+class CloudStorageAzure(models.AbstractModel):
+    _inherit = 'cloud.storage.provider'
+    _description = 'Azure Cloud Storage'
+
+    _url_pattern = re.compile(r'https://(?P<account_name>[\w]+).blob.core.windows.net/(?P<container_name>[\w]+)/(?P<blob_name>[^?]+)')
+
+    def _get_info_from_url(self, url):
+        match = self._url_pattern.match(url)
+        return {
+            'account_name': match.group('account_name'),
+            'container_name': match.group('container_name'),
+            'blob_name': unquote(match.group('blob_name')),
+        }
+
+    def _get_user_delegation_key(self):
+        """ re-generate user_delegation_key every week which won't expire before regeneration """
+        today = date.today()
+        return self._generate_user_delegation_key(today - timedelta(days=today.weekday()))
+
+    @ormcache('unique_key')
+    def _generate_user_delegation_key(self, unique_key):
+        key_start_time = datetime.utcnow()
```

Comment:
https://blog.miguelgrinberg.com/post/it-s-time-for-a-change-datetime-utcnow-is-now-deprecated

Reply:
TIL